File: tests/parity.c
#define AMOS_NO_MAIN
#include "../amos.c"
static void vector(const double *v,int n) {int i;putchar('[');for(i=0;i<n;++i) printf("%s%.17g",i?",":"",v[i]);putchar(']');}
int main(int argc,char **argv)
{
    Snapshot s;unsigned i,n=argc>3?(unsigned)atoi(argv[3]):240;
    uint64_t seed=argc>1?strtoull(argv[1],NULL,10):1;
    int mode=argc>2?atoi(argv[2]):4;
    if(mode) snapshot_sequence(&s,seed,mode);else snapshot_init(&s,seed,0);
    for(i=0;i<n;++i) {
        Transition t;Subject mind;double action=(double)((i*7+i/5)%3)-1.;
        snapshot_step(&s,action,1,0.,&t);
        if(i%13 && i+1!=n) continue;
        mind=s.subject;
        printf("{\"tick\":%" PRIu64 ",\"action\":%.17g,\"predicted\":",t.tick,t.action);vector(t.predicted,3);
        fputs(",\"next\":",stdout);vector(t.next.x,3);
        fputs(",\"h\":",stdout);vector(s.subject.h,AMOS_H);
        fputs(",\"weights\":",stdout);vector(s.subject.weights[2],AMOS_F);
        fputs(",\"body_error\":",stdout);vector(s.subject.body.error,2);
        printf(",\"decision\":%.0f,\"rng\":\"%016" PRIx64 "\",\"support\":%.17g}\n",
            subject_action(&mind,&mind.observation,s.world.target,.1),s.world.rng,t.recognition.support);
    }
    return 0;
}

File: tests/parity.js
const {Amos}=require('../web/amos.js');
const [seed='1',mode='4',steps='240']=process.argv.slice(2);
const a=new Amos(BigInt(seed),Number(mode)!==0?1:0,Number(mode));
for(let i=0;i<Number(steps);i++) {
  const t=a.step((i*7+Math.floor(i/5))%3-1,true,0);
  if(i%13 && i+1!==Number(steps)) continue;
  const s=a.subject,saved=s.rng.state,decision=s.action(a.world.target,.1);s.rng.state=saved;
  console.log(JSON.stringify({tick:t.tick,action:t.action,predicted:t.predicted,next:t.next,
    h:s.h,weights:s.weights[2],body_error:s.body.error,decision,rng:a.world.rng.state.toString(16).padStart(16,'0'),support:s.glyph.enabled?t.recognition.support:0}));
}
